File: Week_01/java/src/_189_RotateArray.java
/**
 * 第一遍：2020/08/14周五 ✅
 * 第二遍：2020/05/22周五
 * 第三遍：2020/05/25周一
 * 第四遍：2020/08/13周四
 * 第五遍：2020/08/14周五
 */
class _189_RotateArray {
    //环状替换
    public void rotate(int[] nums, int k) {
        if (nums == null) return;
        int n = nums.length;
        k = k % n;
        for (int start = 0, count = 0; count < n; start++) {
            int curr = start;
            int pre = nums[start];
            do {
                int next = (curr + k) % n;
                int t = nums[next];
                nums[next] = pre;
                pre = t;
                curr = next;
                count++;
            } while (curr != start);
        }
    }

    class SolutionReverse {
        //反转再反转，和反转字符串中的单词那一题类似。
        public void rotate(int[] nums, int k) {
            if (nums == null) return;
            k = k % nums.length;
            reverse(nums, 0, nums.length - 1);
            reverse(nums, 0, k - 1);
            reverse(nums, k, nums.length - 1);
        }

        void reverse(int[] nums, int l, int r) {
            while (l < r) {
                int t = nums[l];
                nums[l] = nums[r];
                nums[r] = t;
                l++;
                r--;
            }
        }
    }
}

File: Week_01/kotlin/src/_189_RotateArray.kt
fun rotate(nums: IntArray, k: Int): Unit {
    val n = nums.size
    val k_ = k % n
    reverse(nums, 0, n - 1)
    reverse(nums, 0, k_ - 1)
    reverse(nums, k_, n - 1)
}

fun reverse(nums: IntArray, left: Int, right: Int) {
    var l = left
    var r = right
    while (l < r) {
        nums[l] = nums[r].also { nums[r] = nums[l] }
        l++
        r--
    }
}

class SolutionReplace {
    fun rotate(nums: IntArray, k: Int): Unit {
        val n = nums.size
        val k_ = k % n
        var count = 0
        var start = 0
        while (count < n) {
            var curr = start
            var pre = nums[curr]
            do {
                val next = (curr + k_) % n
                nums[next] = pre.also { pre = nums[next] }
                count++
                curr = next
            } while (curr != start)
            start++
        }
    }
}